{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ebf7e24d-89ca-4d9c-b305-fe44df3d0d3b",
   "metadata": {},
   "source": [
    "# A template notebook to run mFinder from Uri Alon's lab\n",
    "\n",
    "- mFinder is a package to compute network motifs, a pattern of motifs that is over-represented than randomly permuted network.\n",
    "\n",
    "- Designed for Windows machine, but can run on Linux environment.\n",
    "- mfinder 1.21 is used here.\n",
    "\n",
    "Last updated: 09/19/2023\n",
    "\n",
    "Author: Yang-Joon Kim\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "92518a16-f5f3-48f7-9a26-249e50e365fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 0. Import\n",
    "import os\n",
    "import sys\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "06dda06e-b6dc-4981-911e-79a246e17970",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yang-joon.kim/.conda/envs/celloracle_env/lib/python3.8/site-packages/loompy/bus_file.py:68: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  def twobit_to_dna(twobit: int, size: int) -> str:\n",
      "/home/yang-joon.kim/.conda/envs/celloracle_env/lib/python3.8/site-packages/loompy/bus_file.py:85: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  def dna_to_twobit(dna: str) -> int:\n",
      "/home/yang-joon.kim/.conda/envs/celloracle_env/lib/python3.8/site-packages/loompy/bus_file.py:102: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  def twobit_1hamming(twobit: int, size: int) -> List[int]:\n",
      "INFO:matplotlib.font_manager:Failed to extract font properties from /usr/share/fonts/google-noto-emoji/NotoColorEmoji.ttf: In FT2Font: Can not load face (unknown file format; error code 0x2)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'0.14.0'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import celloracle as co\n",
    "co.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4659d47a-68d6-426f-ac32-83de542092c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# visualization settings\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "%matplotlib inline\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6, 4.5]\n",
    "plt.rcParams[\"savefig.dpi\"] = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64dd0d29-f48f-4e40-ae88-2cb031be3b07",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bd5c9cba-d179-4f85-a758-0f7339e81e33",
   "metadata": {},
   "source": [
    "# Main Command - using for loop to run mFinder for all cell-types and all timepoints(stages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "81b3a1b1-b2dc-41ec-b735-014cfa3fc96b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change the current working directory\n",
    "#os.chdir(\"/hpc/projects/data.science/yangjoon.kim/github_repos/mfinder/mfinder1.21/\")\n",
    "#! export PATH=/hpc/projects/data.science/yangjoon.kim/github_repos/mfinder/mfinder1.21:$PATH\n",
    "#os.system(\"export PATH=/hpc/projects/data.science/yangjoon.kim/github_repos/mfinder/mfinder1.21:$PATH\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1c7d4fd3-ea78-42c5-b10e-43f29aa87f29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# set the mfinder path in the PATH variable\n",
    "mfinder_path = \"/hpc/projects/data.science/yangjoon.kim/github_repos/mfinder/mfinder1.21\"\n",
    "\n",
    "cmd = \"export PATH=\"+mfinder_path+\":$PATH\"\n",
    "cmd\n",
    "\n",
    "os.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb19244e-35c0-4a5b-b621-b6da6b114ac7",
   "metadata": {},
   "source": [
    "## Step 1. Define the master adata object with all timepoints\n",
    "\n",
    "- "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "76d0d164-fa4d-4812-8211-3ae9d4618422",
   "metadata": {},
   "outputs": [],
   "source": [
    "# move to the directory where the output files will be saved (motif txt)\n",
    "filepath = \"/hpc/projects/data.science/yangjoon.kim/zebrahub_multiome/data/processed_data/baseGRN_CisBP_RNA_zebrahub/09_network_motifs/\"\n",
    "os.chdir(filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "53a3ca69-5b1c-44ef-87d6-d7edd697dddc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 38189 × 27877\n",
       "    obs: 'n_genes', 'n_counts', '10X_run', 'fish', 'fish_n', 'timepoint', 'seqrun', 'log_ncounts', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_nc', 'pct_counts_nc', 'cell_annotation', 'global_annotation'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'nc', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'fish_colors', 'global_annotation_colors', 'hvg', 'neighbors', 'pca', 'timepoint_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the zebrahub early timepoints\n",
    "zebrahub = sc.read_h5ad(\"/hpc/projects/data.science/yangjoon.kim/zebrahub_multiome/data/annotated_data/ZF_atlas_v01/ZF_atlas_v01.h5ad\")\n",
    "zebrahub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d8dd394d-5739-4fbf-be9e-5055076fa2a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['10somite', '0 budstage', '20somite', '30somite', '5somite', '15somite']\n",
       "Categories (6, object): ['0 budstage', '5somite', '10somite', '15somite', '20somite', '30somite']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zebrahub.obs.timepoint.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a5bfa34d-393c-4293-b18b-40cc21ed5f93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['0 budstage', '5somite', '10somite', '15somite', '20somite', '30somite']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "timepoints = ['0 budstage','5somite','10somite','15somite','20somite','30somite']\n",
    "timepoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "451fc64f-0622-4eed-8148-92fb6dfd9793",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Differentiating_Neurons':          source   target  coef_mean  coef_abs             p      -logp\n",
       " 86657     nhlh2      dla   0.143526  0.143526  1.248927e-17  16.903463\n",
       " 34191     foxd2     btg1   0.129191  0.129191  4.572162e-10   9.339878\n",
       " 86689    zbtb18      dla   0.127046  0.127046  1.403359e-13  12.852831\n",
       " 110654  uncx4.1    fgf8a   0.111938  0.111938  4.035099e-13  12.394146\n",
       " 80220   uncx4.1  cyp26c1   0.106434  0.106434  1.181142e-17  16.927698\n",
       " ...         ...      ...        ...       ...           ...        ...\n",
       " 232407    meis3    olig4   0.023177  0.023177  1.894128e-08   7.722591\n",
       " 169224   hoxc1a     irx7  -0.023163  0.023163  4.268921e-08   7.369682\n",
       " 342603    mafbb     tmsb   0.023161  0.023161  7.182542e-09   8.143722\n",
       " 16973     tlx3b    apoc1   0.023159  0.023159  4.092135e-04   3.388050\n",
       " 154290     arxb   hoxb1b  -0.023143  0.023143  7.714583e-09   8.112688\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Endoderm':         source   target  coef_mean  coef_abs             p      -logp\n",
       " 34183    gata5     btg1   0.179640  0.179640  1.108055e-13  12.955439\n",
       " 358464  foxf2a      ved  -0.123337  0.123337  3.921108e-09   8.406591\n",
       " 34213   tcf7l2     btg1   0.111344  0.111344  6.813417e-08   7.166635\n",
       " 34203    foxl1     btg1   0.107486  0.107486  7.240471e-09   8.140233\n",
       " 358444  sox11b      ved   0.093079  0.093079  2.787509e-07   6.554784\n",
       " ...        ...      ...        ...       ...           ...        ...\n",
       " 176058   gata6  kctd15a  -0.015641  0.015641  6.588633e-11  10.181205\n",
       " 180395   jdp2b     krt8   0.015637  0.015637  3.104431e-08   7.508018\n",
       " 379011  foxi3b   znfl2a   0.015636  0.015636  1.096266e-04   3.960084\n",
       " 330690  sox11b     tbx1  -0.015626  0.015626  3.339792e-06   5.476281\n",
       " 7889      shox  akap12b   0.015623  0.015623  2.491926e-06   5.603465\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Epidermal':          source   target  coef_mean  coef_abs             p      -logp\n",
       " 358444   sox11b      ved  -0.283970  0.283970  5.066188e-10   9.295319\n",
       " 34219     gata3     btg1   0.266665  0.266665  6.168458e-13  12.209823\n",
       " 358421   sox11a      ved  -0.263952  0.263952  1.593111e-06   5.797754\n",
       " 245075     hey1     pfn1   0.242987  0.242987  1.083408e-16  15.965208\n",
       " 34199    gata2a     btg1   0.222484  0.222484  1.143262e-09   8.941854\n",
       " ...         ...      ...        ...       ...           ...        ...\n",
       " 289119  phox2bb  shisa2a   0.041243  0.041243  4.173013e-11  10.379550\n",
       " 57957    foxj1b    cldna  -0.041237  0.041237  1.057354e-14  13.975779\n",
       " 57931      uncx    cldna  -0.041220  0.041220  8.402299e-09   8.075602\n",
       " 104658     sox3     eya1   0.041219  0.041219  3.087405e-12  11.510406\n",
       " 316027    msx1b    sox13   0.041218  0.041218  2.356431e-13  12.627745\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Germline':          source      target  coef_mean  coef_abs             p      -logp\n",
       " 358436    sox10         ved  -0.088445  0.088445  2.148496e-11  10.667865\n",
       " 358439    sox4b         ved   0.084442  0.084442  1.429669e-05   4.844765\n",
       " 358441     sox3         ved  -0.063088  0.063088  6.276520e-07   6.202281\n",
       " 154211    mixl1      hoxb1b  -0.062281  0.062281  8.077366e-10   9.092730\n",
       " 358462    mafbb         ved   0.059307  0.059307  3.844072e-04   3.415208\n",
       " ...         ...         ...        ...       ...           ...        ...\n",
       " 17118     rarga       apoc1   0.008696  0.008696  3.102859e-07   6.508238\n",
       " 365066    pitx1  zgc:110425   0.008688  0.008688  8.288355e-06   5.081532\n",
       " 120785    pax3a      fthl27   0.008680  0.008680  1.751482e-07   6.756594\n",
       " 195757  hoxb13a        mdkb   0.008677  0.008677  6.677630e-05   4.175378\n",
       " 161459     sp5a         id3   0.008676  0.008676  5.856458e-05   4.232365\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Lateral_Mesoderm':         source   target  coef_mean  coef_abs             p      -logp\n",
       " 358419  foxc1a      ved  -0.281434  0.281434  1.634938e-11  10.786499\n",
       " 358430   foxd2      ved  -0.235953  0.235953  2.134200e-11  10.670765\n",
       " 358439   sox4b      ved   0.208662  0.208662  9.009097e-09   8.045319\n",
       " 358468    gcm2      ved   0.202189  0.202189  2.252831e-12  11.647271\n",
       " 358443   pax2a      ved  -0.199601  0.199601  1.271333e-11  10.895741\n",
       " ...        ...      ...        ...       ...           ...        ...\n",
       " 87099     drgx      dlc  -0.033177  0.033177  1.086396e-12  11.964012\n",
       " 50118   hoxb7a    cdx1a   0.033176  0.033176  5.435604e-17  16.264752\n",
       " 94465    tbx16  efemp2b   0.033166  0.033166  2.945419e-12  11.530853\n",
       " 314200   foxe3   socs3a   0.033165  0.033165  3.793304e-08   7.420982\n",
       " 271865  mef2ca     rhov   0.033164  0.033164  2.468161e-10   9.607626\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Muscle':         source      target  coef_mean  coef_abs             p      -logp\n",
       " 196972   egr2a      mespab   0.159530  0.159530  3.237219e-16  15.489828\n",
       " 10429   zbtb18      alpi.1   0.142937  0.142937  1.018552e-14  13.992017\n",
       " 332324   mef2d        tbx6   0.138961  0.138961  7.120085e-14  13.147515\n",
       " 2817    zbtb18      actc1a   0.138058  0.138058  1.403663e-11  10.852737\n",
       " 208331  zbtb18       myl10   0.136692  0.136692  5.193597e-14  13.284532\n",
       " ...        ...         ...        ...       ...           ...        ...\n",
       " 272838   foxd2     ripply2   0.029088  0.029088  7.653369e-05   4.116147\n",
       " 365000  dmrt2a  zgc:110425   0.029086  0.029086  2.052247e-08   7.687770\n",
       " 114088    sox6        fn1b   0.029081  0.029081  2.530927e-07   6.596720\n",
       " 375078  hoxc3a   zgc:56585   0.029074  0.029074  1.695497e-12  11.770703\n",
       " 22747    hmx3a        asph  -0.029073  0.029073  1.949037e-04   3.710180\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'NMPs':          source   target  coef_mean  coef_abs             p      -logp\n",
       " 358439    sox4b      ved   0.261958  0.261958  1.277100e-14  13.893775\n",
       " 358422   sox19a      ved  -0.139080  0.139080  4.481563e-13  12.348570\n",
       " 358461   tcf7l2      ved   0.125323  0.125323  1.341573e-08   7.872386\n",
       " 331068    sox4b    tbx16   0.120076  0.120076  1.248914e-15  14.903468\n",
       " 28455      eve1   bambia   0.118482  0.118482  2.315701e-15  14.635317\n",
       " ...         ...      ...        ...       ...           ...        ...\n",
       " 113886     sox2     fn1a  -0.026103  0.026103  1.985295e-10   9.702175\n",
       " 82149    foxj1a     dag1  -0.026098  0.026098  1.041979e-06   5.982141\n",
       " 4693       eve1  adipor2   0.026081  0.026081  7.924082e-17  16.101051\n",
       " 195712  hoxc13a     mdkb   0.026078  0.026078  8.142323e-10   9.089252\n",
       " 153441   hoxa4a   hoxa9a   0.026077  0.026077  7.392506e-10   9.131208\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Neural_Anterior':        source      target  coef_mean  coef_abs             p      -logp\n",
       " 238198   en2a       pax2a   0.268953  0.268953  1.788528e-14  13.747504\n",
       " 77840     rx3      cxcr4a   0.245100  0.245100  8.288253e-16  15.081537\n",
       " 148032   en2a        her5   0.241369  0.241369  7.076461e-18  17.150184\n",
       " 99261   pax2a        en2a   0.228258  0.228258  3.129809e-18  17.504482\n",
       " 368394  foxp4  zgc:153409   0.220184  0.220184  7.511124e-15  14.124295\n",
       " ...       ...         ...        ...       ...           ...        ...\n",
       " 38795   pax3a      calr3b   0.044897  0.044897  1.513212e-13  12.820100\n",
       " 351025   emx2      tuba1a   0.044894  0.044894  1.156708e-07   6.936776\n",
       " 316978   lhx5        sox2   0.044876  0.044876  2.725028e-08   7.564629\n",
       " 23078     jun        atf3   0.044873  0.044873  1.615609e-13  12.791664\n",
       " 372071   en1b  zgc:173587  -0.044860  0.044860  8.466925e-09   8.072274\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Neural_Crest':          source  target  coef_mean  coef_abs             p      -logp\n",
       " 34204      zic6    btg1   0.068596  0.068596  5.324246e-11  10.273742\n",
       " 238415   nr2f1b   pax3a   0.067659  0.067659  1.214941e-16  15.915445\n",
       " 154440    shox2  hoxb3a   0.064674  0.064674  6.621051e-08   7.179073\n",
       " 377656   nr2f1b   zic2a   0.062994  0.062994  3.387868e-09   8.470074\n",
       " 196417    meis3  meis2a   0.061817  0.061817  8.455832e-11  10.072844\n",
       " ...         ...     ...        ...       ...           ...        ...\n",
       " 227304   tfap2a  nrarpa   0.014046  0.014046  1.054752e-08   7.976850\n",
       " 359367  onecut1     vox  -0.014045  0.014045  1.775671e-06   5.750637\n",
       " 240599     sp5l  pcdh19   0.014040  0.014040  2.417189e-09   8.616689\n",
       " 161193    shox2    id2a   0.014038  0.014038  1.756139e-04   3.755441\n",
       " 161192    msx1b    id2a   0.014032  0.014032  1.021096e-08   7.990933\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Neural_Posterior':         source  target  coef_mean  coef_abs             p      -logp\n",
       " 238438   msx1b   pax3a   0.266057  0.266057  2.233909e-17  16.650935\n",
       " 17086    msx1b   apoc1   0.259281  0.259281  1.242339e-18  17.905760\n",
       " 86677   sox11a     dla   0.228103  0.228103  1.201533e-19  18.920264\n",
       " 154172  hnf1ba  hoxb1b   0.200058  0.200058  1.212904e-17  16.916174\n",
       " 152016   meis3  hnf1ba   0.199326  0.199326  2.207879e-17  16.656025\n",
       " ...        ...     ...        ...       ...           ...        ...\n",
       " 378956  hoxc6b  znfl2a   0.041696  0.041696  1.250876e-13  12.902786\n",
       " 195758   klf17    mdkb  -0.041693  0.041693  5.598605e-11  10.251920\n",
       " 110329  tfap2a    fgf3  -0.041686  0.041686  3.707827e-12  11.430881\n",
       " 148220    foxa    her6   0.041684  0.041684  9.189336e-11  10.036716\n",
       " 147824    emx2  her4.4   0.041680  0.041680  1.503864e-09   8.822791\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Notochord':          source      target  coef_mean  coef_abs             p      -logp\n",
       " 34213    tcf7l2        btg1   0.138960  0.138960  1.056249e-11  10.976234\n",
       " 34189    foxi3a        btg1   0.113222  0.113222  1.365006e-08   7.864865\n",
       " 148911    foxd3        hes6   0.107777  0.107777  8.125356e-12  11.090158\n",
       " 113863     mnx1        fn1a   0.106886  0.106886  4.574540e-16  15.339653\n",
       " 34202    dmrt3a        btg1   0.102094  0.102094  9.980856e-13  12.000832\n",
       " ...         ...         ...        ...       ...           ...        ...\n",
       " 161995   sox11a       ier2a   0.020842  0.020842  5.369809e-13  12.270041\n",
       " 368391  phox2bb  zgc:153409  -0.020833  0.020833  2.155462e-06   5.666460\n",
       " 172912    dlx2a       jdp2b   0.020832  0.020832  1.158611e-04   3.936062\n",
       " 147875    nhlh2      her4.4   0.020829  0.020829  1.747587e-09   8.757561\n",
       " 223030   foxj1b       nova2   0.020826  0.020826  2.449883e-07   6.610855\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'PSM':          source      target  coef_mean  coef_abs             p      -logp\n",
       " 358419   foxc1a         ved  -0.369738  0.369738  6.517406e-16  15.185925\n",
       " 358416     pax5         ved   0.251017  0.251017  1.719259e-13  12.764659\n",
       " 358438     egr4         ved   0.199000  0.199000  2.848313e-10   9.545412\n",
       " 358439    sox4b         ved   0.174633  0.174633  1.380857e-07   6.859851\n",
       " 10429    zbtb18      alpi.1   0.156871  0.156871  1.298534e-17  16.886547\n",
       " ...         ...         ...        ...       ...           ...        ...\n",
       " 20530   hoxb10a      arl4aa  -0.031318  0.031318  1.130528e-11  10.946719\n",
       " 227314   foxo1a      nrarpa  -0.031313  0.031313  2.458843e-09   8.609269\n",
       " 346903   hoxc1a       tob1a  -0.031309  0.031309  1.372716e-16  15.862419\n",
       " 165446    pbx1a  im:7138239   0.031307  0.031307  1.128094e-06   5.947655\n",
       " 348245   hoxb5b        tpm3   0.031305  0.031305  6.969616e-13  12.156791\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'Somites':         source   target  coef_mean  coef_abs             p      -logp\n",
       " 358430   foxd2      ved  -0.201453  0.201453  8.021183e-16  15.095762\n",
       " 197037  dmrt2a   mespab   0.198571  0.198571  1.016912e-16  15.992717\n",
       " 196815  hoxc6b   mespaa   0.196897  0.196897  7.465155e-17  16.126961\n",
       " 332389  dmrt2a     tbx6   0.190522  0.190522  5.323086e-15  14.273836\n",
       " 196758  hoxb3a    meox1   0.186918  0.186918  1.006142e-12  11.997341\n",
       " ...        ...      ...        ...       ...           ...        ...\n",
       " 87037    barx1      dlc   0.037743  0.037743  5.479247e-11  10.261279\n",
       " 285163   alx4a  sema3aa   0.037727  0.037727  8.268290e-13  12.082584\n",
       " 22938    tbx16     aspm   0.037720  0.037720  1.547360e-16  15.810409\n",
       " 154452  nr2f1a   hoxb3a   0.037703  0.037703  2.981958e-06   5.525498\n",
       " 156270   six1b   hoxc6b  -0.037702  0.037702  1.789292e-13  12.747319\n",
       " \n",
       " [2000 rows x 6 columns],\n",
       " 'unassigned':         source      target  coef_mean  coef_abs             p      -logp\n",
       " 358439   sox4b         ved   0.312611  0.312611  6.390484e-14  13.194466\n",
       " 180509   gata3       krt91   0.272480  0.272480  1.103619e-12  11.957181\n",
       " 180483  gata2a       krt91   0.262806  0.262806  1.988116e-11  10.701558\n",
       " 358455   sox9b         ved  -0.227327  0.227327  1.463048e-13  12.834741\n",
       " 180474  foxi3a       krt91   0.219864  0.219864  8.605696e-13  12.065214\n",
       " ...        ...         ...        ...       ...           ...        ...\n",
       " 88145    dlx2a       dlx3b   0.053395  0.053395  6.867209e-12  11.163220\n",
       " 203097  hoxb8a       msgn1   0.053377  0.053377  1.636242e-09   8.786153\n",
       " 22891    mef2d        aspm   0.053363  0.053363  3.623438e-13  12.440879\n",
       " 374691    eve1  zgc:194989   0.053357  0.053357  4.533264e-09   8.343589\n",
       " 125811   foxi1      gata2a   0.053347  0.053347  1.572447e-10   9.803424\n",
       " \n",
       " [2000 rows x 6 columns]}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stage = timepoints[0]\n",
    "\n",
    "GRN_stage = co.load_hdf5(links_filepath + \"08_\"+stage+\"_celltype_GRNs.celloracle.links\")\n",
    "GRN_stage.filtered_links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "72ae17c9-875d-4fb1-a374-74136064ec03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/hpc/projects/data.science/yangjoon.kim/zebrahub_multiome/data/processed_data/baseGRN_CisBP_RNA_zebrahub/09_network_motifs/'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filepath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d94b1274-7ce2-4583-87a8-8cbd37e48571",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Differentiating_Neurons\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Differentiating_Neurons network motif computing completed\n",
      "Endoderm\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Endoderm network motif computing completed\n",
      "Epidermal\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Epidermal network motif computing completed\n",
      "Germline\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Germline network motif computing completed\n",
      "Lateral_Mesoderm\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Lateral_Mesoderm network motif computing completed\n",
      "Muscle\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Muscle network motif computing completed\n",
      "NMPs\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_NMPs network motif computing completed\n",
      "Neural_Anterior\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Neural_Anterior network motif computing completed\n",
      "Neural_Crest\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Neural_Crest network motif computing completed\n",
      "Neural_Posterior\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Neural_Posterior network motif computing completed\n",
      "Notochord\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Notochord network motif computing completed\n",
      "PSM\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_PSM network motif computing completed\n",
      "Somites\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_Somites network motif computing completed\n",
      "unassigned\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Usage : mfinder <Network input file name> -s <motif size> -r <no. of randon networks> [-f <output file name>] [more flags]\n",
      "\n",
      "\t-s <motif size>  :Motif size to search\n",
      "\t-r <rand net num> :Number of random networks to generate\n",
      "\t-f <output file name>  : Output file name\n",
      "\t-nd : Input network is a non-directed network.\n",
      "\t-p <num of samples>: run with Sampling method,\n",
      "\t-omem : output members list of all subgraphs\n",
      "\t-h : help\n",
      "\n",
      "\tAdditional flags:\n",
      "\n",
      "\tMotif criteria flags:\n",
      "\t-m <value> : mfactor threshold to use when calculating motifs\n",
      "\t-z <value> : Z-score threshold to use when calculating motifs\n",
      "\t-u : Uniqueness threshold\n",
      "\t-nu : Dont count uniqueness and ignore uniqueness threshold\n",
      "\n",
      "\tRandom networks randomization flags:\n",
      "\n",
      "\t-rs : use stubs method for generating random networks\n",
      "\t-rclust : Preserve clustering sequence in random networks\n",
      "\t-met :Use Metropolis algorithm to conserve triad-census\n",
      "\t\tin random networks\n",
      "\t\t(for s>3; Default : Do not use Metropolis)\n",
      "\t-t0 <(default 0.001)> :Initial temperature (-m0 budstage_unassigned network motif computing completed\n",
      "et option)\n",
      "\t-iter <(default 2)> :controls how many steps to perform (-met option) \n",
      "\t-eth  <(default 0.005)> : energy threshhold (-met option)\n",
      "\t-rgrass <colony size>: generate random networks using grassberger \n",
      "\t\talgorithm\n",
      "\t-rgrass_max_sz <max ratio>: Limit maximal colony size ratio\n",
      "\t-rdm: don't conserve mutual edges in random networks\n",
      "\t-rcl <layers num><size1 size2 ..sizem>: conserve layers in random\n",
      "\t\t networks\n",
      "\t-nsr : Global Switches number when generating Random networks.\n",
      "\n",
      "\n",
      "\tOutput files flags:\n",
      "\n",
      "\t-oi : output intermediate output file. Defualt :No: \n",
      "\t-ospmem <subgraph id>: output members list of a specific subgraphs only\n",
      "\t-maxmem <list length>: limit length of members list to 'list length'.\n",
      "\t\tDefualt: 1000\n",
      "\t-omat : output matrix format file ('__MAT.txt')\n",
      "\t-omet : output metropolis log\n",
      "\t-olog : output general log file\n",
      "\t-orall : output matrix format of appearances in each random network\n",
      "\t-ornet : output random networks files\n",
      "\t-otop <no. of top motifs> : No. of top motifs to show\n",
      "\t-onodangl: output a list of all nan-dangling detected motifs\n",
      "\n",
      "\tOther flags\n",
      "\n",
      "\t-ts : <target,source,weight> Old format of input network file\n",
      "\t-q :Quiet mode - No output to the screen\n",
      "\t-dd : Don't die mode. Wait to user action before terminating the\n",
      "\t\tprogram\n",
      "\t-pold <num of samples>: run sampling method old version\n",
      "\t-nor : Dont search Real network. Defualt :No: \n",
      "\t-cr : calculate roles statistics\n",
      "Differentiating_Neurons\n",
      "Input Network file is filtered_GRN_5somite_Differentiating_Neurons_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Differentiating_Neurons_mfinder_format.txt\n",
      "\tNetwork 5somite_Differentiating_Neurons network motif computing completed\n",
      "Endoderm\n",
      "type: Directed\n",
      "\tNum of Nodes: 421 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 421\n",
      "\tMaximal out degree (out-hub) : 51\n",
      "\tMaximal in degree (in-hub) : 55\n",
      "\tRoots num: 137 Leaves num: 199\n",
      "\tSingle Edges num: 1826 Mutual Edges num: 85\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 34760\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.711+-0.01\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1167\t707.9+-36.3\t12.64\t0.000\t30\t33.57\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t347\t152.6+-14.9\t13.04\t0.000\t12\t9.98\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t174\t47.9+-8.2\t15.38\t0.000\t10\t5.01\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t93\t42.1+-6.9\t7.39\t0.000\t8\t2.68\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t50\t10.9+-2.7\t14.25\t0.000\t7\t1.44\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   32.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Differentiating_Neurons_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Endoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Endoderm_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 5somite_Endoderm network motif computing completed\n",
      "Epidermal\n",
      "457 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 457\n",
      "\tMaximal out degree (out-hub) : 51\n",
      "\tMaximal in degree (in-hub) : 36\n",
      "\tRoots num: 163 Leaves num: 213\n",
      "\tSingle Edges num: 1888 Mutual Edges num: 44\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 31268\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.764+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1081\t502.1+-26.7\t21.67\t0.000\t33\t34.57\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t156\t32.9+-7.3\t16.92\t0.000\t10\t4.99\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t127\t28.7+-5.8\t16.94\t0.000\t8\t4.06\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t49\t12.9+-3.8\t9.51\t0.000\t7\t1.57\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Endoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Epidermal_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Epidermal_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 473 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 473\n",
      "\tMaximal out degree (out-hub) : 53\n",
      "\tMaximal in degree (in-hub) : 51\n",
      "\tRoots num: 140 Leaves num: 246\n",
      "\tSingle Edges num: 1861 Mutual Edges num: 69\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 36099\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.722+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 6 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t876\t689.8+-37.8\t4.93\t0.000\t29\t24.27\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t310\t174.1+-14.8\t9.17\t0.000\t19\t8.59\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "102\t40\t19.8+-4.5\t4.50\t0.000\t11\t1.11\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t43\t17.2+-4.5\t5.77\t0.000\t7\t1.19\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t59\t19.0+-45somite_Epidermal network motif computing completed\n",
      "Lateral_Mesoderm\n",
      ".4\t9.16\t0.000\t8\t1.63\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t15\t3.4+-1.8\t6.47\t0.000\t5\t0.42\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   33.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Epidermal_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Lateral_Mesoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Lateral_Mesoderm_mfinder_format.txt\n",
      "\tNetwork type: Directed5somite_Lateral_Mesoderm network motif computing completed\n",
      "Muscle\n",
      "\n",
      "\tNum of Nodes: 466 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 466\n",
      "\tMaximal out degree (out-hub) : 64\n",
      "\tMaximal in degree (in-hub) : 49\n",
      "\tRoots num: 133 Leaves num: 237\n",
      "\tSingle Edges num: 1873 Mutual Edges num: 61\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 39597\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.710+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1080\t938.9+-44.6\t3.16\t0.000\t29\t27.27\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t288\t141.8+-12.9\t11.29\t0.000\t17\t7.27\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t67\t26.2+-6.0\t6.84\t0.000\t9\t1.69\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t48\t21.9+-5.0\t5.19\t0.000\t6\t1.21\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   35.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Lateral_Mesoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Muscle_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Muscle_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 364 5somite_Muscle network motif computing completed\n",
      "NMPs\n",
      "Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 364\n",
      "\tMaximal out degree (out-hub) : 76\n",
      "\tMaximal in degree (in-hub) : 41\n",
      "\tRoots num: 119 Leaves num: 193\n",
      "\tSingle Edges num: 1876 Mutual Edges num: 61\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 44901\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.603+-0.01\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1730\t1378.9+-62.3\t5.64\t0.000\t22\t38.53\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t570\t306.6+-19.2\t13.69\t0.000\t12\t12.69\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t120\t63.9+-8.7\t6.44\t0.000\t8\t2.67\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t89\t60.7+-9.2\t3.09\t0.000\t5\t1.98\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t39\t18.3+-2.8\t7.27\t0.000\t4\t0.87\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   36.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Muscle_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_NMPs_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_NMPs_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 390 Num 5somite_NMPs network motif computing completed\n",
      "Neural_Anterior\n",
      "of Edges: 2000\n",
      "\tNum of Nodes with edges: 390\n",
      "\tMaximal out degree (out-hub) : 70\n",
      "\tMaximal in degree (in-hub) : 44\n",
      "\tRoots num: 120 Leaves num: 201\n",
      "\tSingle Edges num: 1871 Mutual Edges num: 64\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 43797\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.625+-0.01\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1580\t1209.4+-52.0\t7.12\t0.000\t25\t36.08\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t611\t382.2+-22.6\t10.12\t0.000\t14\t13.95\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t118\t62.8+-8.0\t6.86\t0.000\t9\t2.69\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t17\t7.8+-2.1\t4.37\t0.000\t5\t0.39\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   36.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_5somite_NMPs_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Neural_Anterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Neural_Anterior_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\t5somite_Neural_Anterior network motif computing completed\n",
      "Neural_Crest\n",
      "Num of Nodes: 462 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 462\n",
      "\tMaximal out degree (out-hub) : 46\n",
      "\tMaximal in degree (in-hub) : 62\n",
      "\tRoots num: 134 Leaves num: 222\n",
      "\tSingle Edges num: 1841 Mutual Edges num: 79\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 33835\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.763+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 7 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t745\t582.7+-30.3\t5.35\t0.000\t34\t22.02\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t269\t131.4+-12.2\t11.28\t0.000\t21\t7.95\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "98\t13\t6.8+-2.8\t2.17\t0.030\t4\t0.38\n",
      "\n",
      "0 1 0 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "102\t30\t16.3+-4.4\t3.11\t0.000\t10\t0.89\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t59\t12.6+-3.5\t13.20\t0.000\t12\t1.74\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t36\t12.1+-3.4\t7.05\t0.000\t9\t1.06\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t9\t1.0+-0.9\t8.62\t0.000\t7\t0.27\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   33.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Neural_Anterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Neural_Crest_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Neural_Crest_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of5somite_Neural_Crest network motif computing completed\n",
      "Neural_Posterior\n",
      " Nodes: 362 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 362\n",
      "\tMaximal out degree (out-hub) : 71\n",
      "\tMaximal in degree (in-hub) : 48\n",
      "\tRoots num: 124 Leaves num: 179\n",
      "\tSingle Edges num: 1853 Mutual Edges num: 73\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 42436\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.628+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1585\t1315.5+-49.2\t5.48\t0.000\t28\t37.35\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t384\t260.2+-18.3\t6.76\t0.000\t13\t9.05\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t158\t114.3+-12.1\t3.62\t0.000\t9\t3.72\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t93\t67.2+-7.8\t3.29\t0.000\t7\t2.19\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t21\t9.5+-2.6\t4.51\t0.000\t5\t0.49\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   35.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Neural_Crest_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Neural_Posterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Neural_Posterior_mfinder_format.txt\n",
      "\tNetwork type: Directed5somite_Neural_Posterior network motif computing completed\n",
      "Notochord\n",
      "\n",
      "\tNum of Nodes: 471 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 471\n",
      "\tMaximal out degree (out-hub) : 45\n",
      "\tMaximal in degree (in-hub) : 55\n",
      "\tRoots num: 110 Leaves num: 256\n",
      "\tSingle Edges num: 1823 Mutual Edges num: 88\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 33736\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.765+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 6 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t813\t618.0+-31.6\t6.16\t0.000\t30\t24.10\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t301\t153.7+-13.3\t11.12\t0.000\t19\t8.92\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "102\t29\t20.4+-4.1\t2.08\t0.020\t9\t0.86\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t35\t12.6+-3.2\t6.93\t0.000\t11\t1.04\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t48\t14.8+-3.6\t9.29\t0.000\t11\t1.42\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t11\t1.8+-1.2\t7.72\t0.000\t5\t0.33\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   33.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Neural_Posterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Notochord_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Notochord_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes5somite_Notochord network motif computing completed\n",
      "PSM\n",
      ": 462 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 462\n",
      "\tMaximal out degree (out-hub) : 97\n",
      "\tMaximal in degree (in-hub) : 29\n",
      "\tRoots num: 132 Leaves num: 258\n",
      "\tSingle Edges num: 1908 Mutual Edges num: 45\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 39170\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.707+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1439\t723.6+-42.9\t16.69\t0.000\t31\t36.74\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t492\t158.4+-15.1\t22.11\t0.000\t12\t12.56\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t79\t27.1+-5.5\t9.48\t0.000\t11\t2.02\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t57\t22.2+-4.9\t7.13\t0.000\t5\t1.46\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   34.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Notochord_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_PSM_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_PSM_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 424 Num of5somite_PSM network motif computing completed\n",
      "Somites\n",
      " Edges: 2000\n",
      "\tNum of Nodes with edges: 424\n",
      "\tMaximal out degree (out-hub) : 86\n",
      "\tMaximal in degree (in-hub) : 66\n",
      "\tRoots num: 162 Leaves num: 212\n",
      "\tSingle Edges num: 1908 Mutual Edges num: 44\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 51652\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.570+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t408\t328.3+-18.7\t4.27\t0.000\t12\t7.90\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "110\t35\t24.4+-4.2\t2.55\t0.010\t6\t0.68\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   35.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_PSM_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_Somites_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_Somites_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 385somite_Somites network motif computing completed\n",
      "unassigned\n",
      "4 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 384\n",
      "\tMaximal out degree (out-hub) : 76\n",
      "\tMaximal in degree (in-hub) : 49\n",
      "\tRoots num: 125 Leaves num: 204\n",
      "\tSingle Edges num: 1873 Mutual Edges num: 60\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 46540\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.599+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t525\t364.1+-19.9\t8.09\t0.000\t13\t11.28\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t59\t35.1+-5.9\t4.06\t0.000\t6\t1.27\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t70\t44.2+-7.3\t3.52\t0.000\t6\t1.50\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t23\t11.1+-2.9\t4.14\t0.000\t4\t0.49\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   36.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_5somite_Somites_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_5somite_unassigned_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_5somite_unassigned_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nod5somite_unassigned network motif computing completed\n",
      "es: 497 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 497\n",
      "\tMaximal out degree (out-hub) : 52\n",
      "\tMaximal in degree (in-hub) : 46\n",
      "\tRoots num: 178 Leaves num: 228\n",
      "\tSingle Edges num: 1873 Mutual Edges num: 62\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 27999\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.818+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 6 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t973\t405.6+-28.6\t19.83\t0.000\t36\t34.75\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t228\t54.0+-8.3\t21.02\t0.000\t17\t8.14\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "102\t29\t14.3+-4.2\t3.53\t0.010\t9\t1.04\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t96\t20.2+-4.6\t16.52\t0.000\t13\t3.43\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t44\t11.1+-3.0\t11.03\t0.000\t8\t1.57\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t9\t1.4+-1.1\t7.15\t0.000\t5\t0.32\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   29.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_5somite_unassigned_OUT.txt was generated\n",
      "Differentiating_Neurons\n",
      "Input Network file is filtered_GRN_10somite_Differentiating_Neurons_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Differentiating_Neurons_mfinder_format.txt\n",
      "\tNetwor10somite_Differentiating_Neurons network motif computing completed\n",
      "Endoderm\n",
      "k type: Directed\n",
      "\tNum of Nodes: 432 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 432\n",
      "\tMaximal out degree (out-hub) : 49\n",
      "\tMaximal in degree (in-hub) : 38\n",
      "\tRoots num: 117 Leaves num: 224\n",
      "\tSingle Edges num: 1864 Mutual Edges num: 67\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 32215\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.746+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 6 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1073\t536.0+-31.5\t17.04\t0.000\t33\t33.31\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t337\t91.6+-11.2\t22.01\t0.000\t15\t10.46\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "102\t22\t13.4+-4.0\t2.16\t0.020\t8\t0.68\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t108\t23.5+-5.0\t17.04\t0.000\t7\t3.35\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t63\t21.4+-4.3\t9.71\t0.000\t7\t1.96\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t38\t6.8+-2.4\t13.04\t0.000\t4\t1.18\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Differentiating_Neurons_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Endoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Endoderm_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes10somite_Endoderm network motif computing completed\n",
      "Epidermal\n",
      ": 420 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 420\n",
      "\tMaximal out degree (out-hub) : 75\n",
      "\tMaximal in degree (in-hub) : 32\n",
      "\tRoots num: 98 Leaves num: 230\n",
      "\tSingle Edges num: 1909 Mutual Edges num: 45\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 38429\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.708+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1366\t816.5+-41.4\t13.27\t0.000\t29\t35.55\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t256\t85.1+-11.6\t14.77\t0.000\t10\t6.66\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "102\t38\t20.9+-5.2\t3.29\t0.000\t6\t0.99\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t118\t39.6+-6.1\t12.84\t0.000\t9\t3.07\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t46\t16.6+-3.9\t7.47\t0.000\t5\t1.20\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   33.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Endoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Epidermal_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      8 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Epidermal_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nod10somite_Epidermal network motif computing completed\n",
      "Lateral_Mesoderm\n",
      "es: 598 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 598\n",
      "\tMaximal out degree (out-hub) : 152\n",
      "\tMaximal in degree (in-hub) : 25\n",
      "\tRoots num: 108 Leaves num: 412\n",
      "\tSingle Edges num: 1936 Mutual Edges num: 31\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 50424\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.722+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t82\t23.7+-5.3\t11.09\t0.000\t11\t1.63\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t13\t4.5+-2.2\t3.94\t0.000\t6\t0.26\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t7\t0.7+-0.8\t8.23\t0.000\t5\t0.14\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   41.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Epidermal_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Lateral_Mesoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Lateral_Mesoderm_mfinder_format.txt\n",
      "\tNetwork type: Direct10somite_Lateral_Mesoderm network motif computing completed\n",
      "Muscle\n",
      "ed\n",
      "\tNum of Nodes: 507 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 507\n",
      "\tMaximal out degree (out-hub) : 61\n",
      "\tMaximal in degree (in-hub) : 41\n",
      "\tRoots num: 129 Leaves num: 275\n",
      "\tSingle Edges num: 1912 Mutual Edges num: 43\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 31708\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.804+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t931\t504.4+-28.0\t15.24\t0.000\t36\t29.36\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t157\t63.5+-9.1\t10.29\t0.000\t16\t4.95\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t32\t9.5+-3.2\t7.10\t0.000\t11\t1.01\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t16\t4.7+-2.2\t5.22\t0.000\t5\t0.50\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Lateral_Mesoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Muscle_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     32 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     22 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      8 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Muscle_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 3810somite_Muscle network motif computing completed\n",
      "NMPs\n",
      "9 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 389\n",
      "\tMaximal out degree (out-hub) : 116\n",
      "\tMaximal in degree (in-hub) : 39\n",
      "\tRoots num: 85 Leaves num: 248\n",
      "\tSingle Edges num: 1903 Mutual Edges num: 47\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 53508\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.572+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1811\t1624.8+-74.5\t2.50\t0.010\t22\t33.85\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t512\t336.8+-22.8\t7.69\t0.000\t11\t9.57\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   39.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Muscle_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_NMPs_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_NMPs_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 395 Nu10somite_NMPs network motif computing completed\n",
      "Neural_Anterior\n",
      "m of Edges: 2000\n",
      "\tNum of Nodes with edges: 395\n",
      "\tMaximal out degree (out-hub) : 86\n",
      "\tMaximal in degree (in-hub) : 46\n",
      "\tRoots num: 87 Leaves num: 228\n",
      "\tSingle Edges num: 1905 Mutual Edges num: 47\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 46376\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.605+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1995\t1301.3+-66.1\t10.50\t0.000\t29\t43.02\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t505\t237.7+-22.4\t11.93\t0.000\t12\t10.89\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t124\t60.1+-8.3\t7.69\t0.000\t9\t2.67\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t44\t25.5+-4.3\t4.26\t0.000\t6\t0.95\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t9\t1.9+-1.4\t5.23\t0.000\t6\t0.19\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   35.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_NMPs_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Neural_Anterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Neural_Anterior_mfinder_format.txt\n",
      "\tNetwork type: Directed10somite_Neural_Anterior network motif computing completed\n",
      "Neural_Crest\n",
      "\n",
      "\tNum of Nodes: 488 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 488\n",
      "\tMaximal out degree (out-hub) : 65\n",
      "\tMaximal in degree (in-hub) : 51\n",
      "\tRoots num: 140 Leaves num: 240\n",
      "\tSingle Edges num: 1858 Mutual Edges num: 70\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 32105\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.774+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t704\t483.6+-26.0\t8.46\t0.000\t35\t21.93\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t298\t118.5+-11.0\t16.27\t0.000\t23\t9.28\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t64\t12.1+-3.3\t15.83\t0.000\t10\t1.99\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t38\t9.6+-3.1\t9.27\t0.000\t12\t1.18\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t12\t1.1+-1.1\t10.17\t0.000\t6\t0.37\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Neural_Anterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Neural_Crest_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Neural_Crest_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum 10somite_Neural_Crest network motif computing completed\n",
      "Neural_Posterior\n",
      "of Nodes: 350 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 350\n",
      "\tMaximal out degree (out-hub) : 79\n",
      "\tMaximal in degree (in-hub) : 35\n",
      "\tRoots num: 96 Leaves num: 180\n",
      "\tSingle Edges num: 1843 Mutual Edges num: 78\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 43390\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.623+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1737\t1384.0+-59.4\t5.94\t0.000\t26\t40.03\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t596\t338.0+-18.2\t14.21\t0.000\t17\t13.74\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t195\t103.3+-11.0\t8.32\t0.000\t9\t4.49\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t106\t80.1+-8.8\t2.94\t0.000\t9\t2.44\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t24\t11.2+-3.3\t3.86\t0.000\t5\t0.55\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   35.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Neural_Crest_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Neural_Posterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Neural_Posterior_mfinder_format.txt\n",
      "\tNetwork type: Direct10somite_Neural_Posterior network motif computing completed\n",
      "Notochord\n",
      "ed\n",
      "\tNum of Nodes: 482 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 482\n",
      "\tMaximal out degree (out-hub) : 42\n",
      "\tMaximal in degree (in-hub) : 64\n",
      "\tRoots num: 140 Leaves num: 242\n",
      "\tSingle Edges num: 1857 Mutual Edges num: 71\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 33396\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.764+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 6 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t690\t542.4+-29.8\t4.95\t0.000\t32\t20.66\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t302\t143.5+-11.8\t13.40\t0.000\t21\t9.04\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "102\t31\t12.9+-3.7\t4.89\t0.000\t7\t0.93\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t49\t9.5+-3.2\t12.55\t0.000\t7\t1.47\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t37\t10.4+-2.9\t9.30\t0.000\t8\t1.11\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t7\t1.3+-1.2\t4.81\t0.000\t5\t0.21\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Neural_Posterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Notochord_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Notochord_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nod10somite_Notochord network motif computing completed\n",
      "PSM\n",
      "es: 437 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 437\n",
      "\tMaximal out degree (out-hub) : 113\n",
      "\tMaximal in degree (in-hub) : 35\n",
      "\tRoots num: 114 Leaves num: 251\n",
      "\tSingle Edges num: 1908 Mutual Edges num: 43\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 40883\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.669+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1163\t616.9+-37.3\t14.62\t0.000\t29\t28.45\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t319\t76.8+-15.3\t15.81\t0.000\t11\t7.80\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t187\t50.8+-7.6\t18.02\t0.000\t11\t4.57\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t51\t20.3+-4.8\t6.37\t0.000\t4\t1.25\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t20\t6.2+-1.9\t7.32\t0.000\t5\t0.49\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   34.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Notochord_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_PSM_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     32 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     22 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      8 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_PSM_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 360 Num 10somite_PSM network motif computing completed\n",
      "Somites\n",
      "of Edges: 2000\n",
      "\tNum of Nodes with edges: 360\n",
      "\tMaximal out degree (out-hub) : 93\n",
      "\tMaximal in degree (in-hub) : 46\n",
      "\tRoots num: 109 Leaves num: 194\n",
      "\tSingle Edges num: 1893 Mutual Edges num: 53\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 55242\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.522+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t714\t559.4+-23.3\t6.63\t0.000\t11\t12.92\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t57\t40.3+-7.5\t2.23\t0.010\t5\t1.03\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   38.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_10somite_PSM_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_Somites_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     22 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      8 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_Somites_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 10somite_Somites network motif computing completed\n",
      "unassigned\n",
      "378 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 378\n",
      "\tMaximal out degree (out-hub) : 109\n",
      "\tMaximal in degree (in-hub) : 52\n",
      "\tRoots num: 109 Leaves num: 211\n",
      "\tSingle Edges num: 1889 Mutual Edges num: 54\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 51788\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.576+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1788\t1601.4+-60.5\t3.09\t0.010\t22\t34.53\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t596\t384.1+-23.1\t9.17\t0.000\t14\t11.51\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t84\t60.5+-9.7\t2.43\t0.000\t9\t1.62\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t10\t4.1+-1.9\t3.08\t0.000\t4\t0.19\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   38.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_10somite_Somites_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_10somite_unassigned_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_10somite_unassigned_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of N10somite_unassigned network motif computing completed\n",
      "odes: 459 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 459\n",
      "\tMaximal out degree (out-hub) : 44\n",
      "\tMaximal in degree (in-hub) : 32\n",
      "\tRoots num: 142 Leaves num: 233\n",
      "\tSingle Edges num: 1911 Mutual Edges num: 44\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 31684\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.772+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1393\t579.6+-34.6\t23.50\t0.000\t35\t43.97\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t326\t67.5+-8.0\t32.35\t0.000\t13\t10.29\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t90\t25.4+-4.9\t13.30\t0.000\t14\t2.84\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t29\t10.0+-2.8\t6.78\t0.000\t7\t0.92\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   32.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_10somite_unassigned_OUT.txt was generated\n",
      "Adaxial_Cells\n",
      "Input Network file is filtered_GRN_15somite_Adaxial_Cells_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     22 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      8 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Adaxial_Cells_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNu15somite_Adaxial_Cells network motif computing completed\n",
      "Differentiating_Neurons\n",
      "m of Nodes: 465 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 465\n",
      "\tMaximal out degree (out-hub) : 126\n",
      "\tMaximal in degree (in-hub) : 36\n",
      "\tRoots num: 140 Leaves num: 287\n",
      "\tSingle Edges num: 1957 Mutual Edges num: 21\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 49015\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.637+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1118\t724.7+-45.2\t8.70\t0.000\t18\t22.81\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t314\t171.7+-17.9\t7.96\t0.000\t7\t6.41\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t34\t12.9+-3.5\t6.05\t0.000\t5\t0.69\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   38.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Adaxial_Cells_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Differentiating_Neurons_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Differentiating_Neurons_mfinder_format.txt\n",
      "\tNetwor15somite_Differentiating_Neurons network motif computing completed\n",
      "Endoderm\n",
      "k type: Directed\n",
      "\tNum of Nodes: 496 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 496\n",
      "\tMaximal out degree (out-hub) : 79\n",
      "\tMaximal in degree (in-hub) : 48\n",
      "\tRoots num: 154 Leaves num: 271\n",
      "\tSingle Edges num: 1941 Mutual Edges num: 29\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 35969\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.733+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t726\t412.0+-32.4\t9.69\t0.000\t26\t20.18\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t154\t26.7+-6.9\t18.37\t0.000\t13\t4.28\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t43\t5.8+-2.5\t15.15\t0.000\t7\t1.20\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t25\t1.8+-1.2\t19.24\t0.000\t5\t0.70\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   30.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Differentiating_Neurons_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Endoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Endoderm_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes15somite_Endoderm network motif computing completed\n",
      "Epidermal\n",
      ": 483 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 483\n",
      "\tMaximal out degree (out-hub) : 59\n",
      "\tMaximal in degree (in-hub) : 36\n",
      "\tRoots num: 150 Leaves num: 252\n",
      "\tSingle Edges num: 1954 Mutual Edges num: 21\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 32744\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.769+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t768\t420.6+-25.2\t13.78\t0.000\t30\t23.45\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t72\t17.1+-4.2\t13.12\t0.000\t10\t2.20\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t25\t4.6+-2.2\t9.29\t0.000\t8\t0.76\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t8\t0.7+-0.8\t8.62\t0.000\t4\t0.24\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Endoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Epidermal_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Epidermal_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nod15somite_Epidermal network motif computing completed\n",
      "Lateral_Mesoderm\n",
      "es: 530 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 530\n",
      "\tMaximal out degree (out-hub) : 79\n",
      "\tMaximal in degree (in-hub) : 73\n",
      "\tRoots num: 184 Leaves num: 280\n",
      "\tSingle Edges num: 1956 Mutual Edges num: 21\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 46690\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.605+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 1 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t183\t87.6+-10.4\t9.19\t0.000\t8\t3.92\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Epidermal_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Lateral_Mesoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Lateral_Mesoderm_mfinder_format.txt\n",
      "\tNetwork type: Direct15somite_Lateral_Mesoderm network motif computing completed\n",
      "Muscle\n",
      "ed\n",
      "\tNum of Nodes: 593 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 593\n",
      "\tMaximal out degree (out-hub) : 51\n",
      "\tMaximal in degree (in-hub) : 39\n",
      "\tRoots num: 164 Leaves num: 343\n",
      "\tSingle Edges num: 1954 Mutual Edges num: 21\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 28629\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.834+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t506\t300.0+-25.2\t8.17\t0.000\t31\t17.67\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t50\t21.7+-4.8\t5.93\t0.000\t12\t1.75\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t14\t2.4+-1.6\t7.06\t0.000\t6\t0.49\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   29.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Lateral_Mesoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Muscle_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Muscle_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 4715somite_Muscle network motif computing completed\n",
      "NMPs\n",
      "7 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 477\n",
      "\tMaximal out degree (out-hub) : 94\n",
      "\tMaximal in degree (in-hub) : 31\n",
      "\tRoots num: 143 Leaves num: 295\n",
      "\tSingle Edges num: 1958 Mutual Edges num: 16\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 47396\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.662+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1081\t743.7+-52.5\t6.42\t0.000\t16\t22.81\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t191\t91.2+-11.8\t8.45\t0.000\t6\t4.03\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t26\t13.8+-3.6\t3.42\t0.000\t4\t0.55\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   37.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Muscle_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_NMPs_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_NMPs_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 449 Nu15somite_NMPs network motif computing completed\n",
      "Neural_Anterior\n",
      "m of Edges: 2000\n",
      "\tNum of Nodes with edges: 449\n",
      "\tMaximal out degree (out-hub) : 74\n",
      "\tMaximal in degree (in-hub) : 42\n",
      "\tRoots num: 162 Leaves num: 228\n",
      "\tSingle Edges num: 1961 Mutual Edges num: 19\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 38397\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.684+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1013\t566.4+-42.5\t10.51\t0.000\t22\t26.38\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t60\t17.7+-4.5\t9.39\t0.000\t8\t1.56\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t33\t6.9+-2.8\t9.42\t0.000\t7\t0.86\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   32.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_NMPs_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Neural_Anterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Neural_Anterior_mfinder_format.txt\n",
      "\tNetwork type: Directed15somite_Neural_Anterior network motif computing completed\n",
      "Neural_Crest\n",
      "\n",
      "\tNum of Nodes: 519 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 519\n",
      "\tMaximal out degree (out-hub) : 56\n",
      "\tMaximal in degree (in-hub) : 63\n",
      "\tRoots num: 167 Leaves num: 272\n",
      "\tSingle Edges num: 1911 Mutual Edges num: 43\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 36057\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.722+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t530\t411.4+-27.4\t4.32\t0.000\t25\t14.70\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t163\t93.3+-9.1\t7.68\t0.000\t14\t4.52\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "102\t11\t4.1+-2.2\t3.09\t0.010\t4\t0.31\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t16\t5.4+-2.0\t5.23\t0.000\t6\t0.44\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t14\t3.8+-1.8\t5.64\t0.000\t5\t0.39\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   30.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Neural_Anterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Neural_Crest_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Neural_Crest_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum 15somite_Neural_Crest network motif computing completed\n",
      "Neural_Posterior\n",
      "of Nodes: 463 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 463\n",
      "\tMaximal out degree (out-hub) : 93\n",
      "\tMaximal in degree (in-hub) : 56\n",
      "\tRoots num: 168 Leaves num: 235\n",
      "\tSingle Edges num: 1931 Mutual Edges num: 34\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 38035\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.695+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t707\t515.1+-36.7\t5.22\t0.000\t22\t18.59\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t201\t76.9+-9.3\t13.41\t0.000\t11\t5.28\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "98\t5\t1.4+-1.3\t2.85\t0.030\t4\t0.13\n",
      "\n",
      "0 1 0 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t34\t9.5+-2.9\t8.37\t0.000\t8\t0.89\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t26\t5.6+-2.3\t9.07\t0.000\t6\t0.68\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   32.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Neural_Crest_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Neural_Posterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Neural_Posterior_mfinder_format.txt\n",
      "\tNetwork type: Direct15somite_Neural_Posterior network motif computing completed\n",
      "Notochord\n",
      "ed\n",
      "\tNum of Nodes: 532 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 532\n",
      "\tMaximal out degree (out-hub) : 86\n",
      "\tMaximal in degree (in-hub) : 52\n",
      "\tRoots num: 167 Leaves num: 287\n",
      "\tSingle Edges num: 1931 Mutual Edges num: 34\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 34809\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.746+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t686\t395.5+-29.5\t9.86\t0.000\t29\t19.71\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t173\t46.5+-7.7\t16.45\t0.000\t13\t4.97\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t21\t3.4+-1.9\t9.45\t0.000\t7\t0.60\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t16\t1.9+-1.4\t10.15\t0.000\t7\t0.46\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Neural_Posterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Notochord_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Notochord_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nod15somite_Notochord network motif computing completed\n",
      "PSM\n",
      "es: 564 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 564\n",
      "\tMaximal out degree (out-hub) : 78\n",
      "\tMaximal in degree (in-hub) : 36\n",
      "\tRoots num: 170 Leaves num: 340\n",
      "\tSingle Edges num: 1979 Mutual Edges num: 10\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 35363\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.766+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t741\t305.9+-29.7\t14.67\t0.000\t22\t20.95\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t87\t19.7+-5.6\t11.94\t0.000\t7\t2.46\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   32.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Notochord_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_PSM_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_PSM_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 462 Num 15somite_PSM network motif computing completed\n",
      "Somites\n",
      "of Edges: 2000\n",
      "\tNum of Nodes with edges: 462\n",
      "\tMaximal out degree (out-hub) : 84\n",
      "\tMaximal in degree (in-hub) : 61\n",
      "\tRoots num: 181 Leaves num: 236\n",
      "\tSingle Edges num: 1957 Mutual Edges num: 20\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 44941\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.665+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1099\t962.3+-44.0\t3.11\t0.000\t23\t24.45\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t156\t99.8+-12.6\t4.45\t0.000\t8\t3.47\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t27\t8.0+-2.8\t6.87\t0.000\t4\t0.60\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   34.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_PSM_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_Somites_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_Somites_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 15somite_Somites network motif computing completed\n",
      "unassigned\n",
      "429 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 429\n",
      "\tMaximal out degree (out-hub) : 86\n",
      "\tMaximal in degree (in-hub) : 61\n",
      "\tRoots num: 143 Leaves num: 224\n",
      "\tSingle Edges num: 1927 Mutual Edges num: 36\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 45352\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.641+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t322\t219.5+-15.4\t6.66\t0.000\t15\t7.10\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t32\t11.1+-3.9\t5.38\t0.000\t5\t0.71\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   36.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_15somite_Somites_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_15somite_unassigned_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_15somite_unassigned_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of N15somite_unassigned network motif computing completed\n",
      "odes: 501 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 501\n",
      "\tMaximal out degree (out-hub) : 61\n",
      "\tMaximal in degree (in-hub) : 51\n",
      "\tRoots num: 178 Leaves num: 271\n",
      "\tSingle Edges num: 1957 Mutual Edges num: 19\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 36255\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.731+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t845\t523.6+-35.1\t9.14\t0.000\t26\t23.31\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t112\t40.9+-8.6\t8.31\t0.000\t8\t3.09\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t41\t10.2+-3.3\t9.20\t0.000\t6\t1.13\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_15somite_unassigned_OUT.txt was generated\n",
      "Differentiating_Neurons\n",
      "Input Network file is filtered_GRN_20somite_Differentiating_Neurons_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Differentiating_Neurons_mfinder_format.txt\n",
      "\tNetwor20somite_Differentiating_Neurons network motif computing completed\n",
      "Endoderm\n",
      "k type: Directed\n",
      "\tNum of Nodes: 436 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 436\n",
      "\tMaximal out degree (out-hub) : 59\n",
      "\tMaximal in degree (in-hub) : 48\n",
      "\tRoots num: 131 Leaves num: 226\n",
      "\tSingle Edges num: 1858 Mutual Edges num: 69\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 36304\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.702+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 5 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1353\t797.5+-37.3\t14.90\t0.000\t30\t37.27\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t435\t132.6+-13.8\t21.88\t0.000\t14\t11.98\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t123\t37.2+-5.9\t14.64\t0.000\t10\t3.39\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t90\t26.8+-5.1\t12.29\t0.000\t8\t2.48\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "238\t24\t5.2+-2.1\t8.93\t0.000\t4\t0.66\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 1 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Differentiating_Neurons_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Endoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     32 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     30 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     22 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      5 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Endoderm_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes20somite_Endoderm network motif computing completed\n",
      "Epidermal\n",
      ": 458 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 458\n",
      "\tMaximal out degree (out-hub) : 191\n",
      "\tMaximal in degree (in-hub) : 32\n",
      "\tRoots num: 85 Leaves num: 325\n",
      "\tSingle Edges num: 1966 Mutual Edges num: 15\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 67181\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.556+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t96\t65.5+-8.6\t3.52\t0.010\t4\t1.43\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t12\t6.4+-2.7\t2.05\t0.020\t6\t0.18\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   47.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.5 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Endoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Epidermal_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Epidermal_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nod20somite_Epidermal network motif computing completed\n",
      "Lateral_Mesoderm\n",
      "es: 549 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 549\n",
      "\tMaximal out degree (out-hub) : 104\n",
      "\tMaximal in degree (in-hub) : 32\n",
      "\tRoots num: 142 Leaves num: 327\n",
      "\tSingle Edges num: 1941 Mutual Edges num: 29\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 41564\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.733+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t636\t549.4+-36.8\t2.35\t0.020\t23\t15.30\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t175\t64.1+-11.3\t9.82\t0.000\t13\t4.21\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t11\t2.8+-1.6\t5.00\t0.000\t6\t0.26\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t6\t1.3+-1.1\t4.46\t0.000\t4\t0.14\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   35.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Epidermal_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Lateral_Mesoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Lateral_Mesoderm_mfinder_format.txt\n",
      "\tNetwork type: Direct20somite_Lateral_Mesoderm network motif computing completed\n",
      "Muscle\n",
      "ed\n",
      "\tNum of Nodes: 602 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 602\n",
      "\tMaximal out degree (out-hub) : 42\n",
      "\tMaximal in degree (in-hub) : 53\n",
      "\tRoots num: 169 Leaves num: 349\n",
      "\tSingle Edges num: 1934 Mutual Edges num: 27\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 26679\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.843+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 6 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t669\t265.4+-19.3\t20.92\t0.000\t35\t25.08\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t154\t26.8+-5.5\t23.15\t0.000\t11\t5.77\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "98\t6\t1.9+-1.5\t2.67\t0.020\t5\t0.22\n",
      "\n",
      "0 1 0 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "102\t9\t3.6+-1.9\t2.92\t0.010\t4\t0.34\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t25\t5.3+-2.5\t7.91\t0.000\t10\t0.94\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t12\t1.3+-1.0\t10.38\t0.000\t5\t0.45\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   29.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Lateral_Mesoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Muscle_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     22 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      8 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Muscle_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 4120somite_Muscle network motif computing completed\n",
      "NMPs\n",
      "8 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 418\n",
      "\tMaximal out degree (out-hub) : 114\n",
      "\tMaximal in degree (in-hub) : 30\n",
      "\tRoots num: 91 Leaves num: 270\n",
      "\tSingle Edges num: 1937 Mutual Edges num: 31\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 51494\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.619+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1899\t1186.8+-68.6\t10.38\t0.000\t22\t36.88\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t392\t208.9+-22.2\t8.23\t0.000\t9\t7.61\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t71\t31.7+-6.0\t6.53\t0.000\t7\t1.38\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   40.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Muscle_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_NMPs_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_NMPs_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 468 Nu20somite_NMPs network motif computing completed\n",
      "Neural_Anterior\n",
      "m of Edges: 2000\n",
      "\tNum of Nodes with edges: 468\n",
      "\tMaximal out degree (out-hub) : 88\n",
      "\tMaximal in degree (in-hub) : 30\n",
      "\tRoots num: 121 Leaves num: 282\n",
      "\tSingle Edges num: 1959 Mutual Edges num: 20\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 43431\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.666+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1549\t752.0+-46.1\t17.29\t0.000\t25\t35.67\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t270\t102.1+-13.7\t12.23\t0.000\t5\t6.22\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t43\t23.6+-4.9\t3.97\t0.000\t6\t0.99\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   35.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_20somite_NMPs_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Neural_Anterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Neural_Anterior_mfinder_format.txt\n",
      "\tNetwork type: Directed20somite_Neural_Anterior network motif computing completed\n",
      "Neural_Crest\n",
      "\n",
      "\tNum of Nodes: 463 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 463\n",
      "\tMaximal out degree (out-hub) : 78\n",
      "\tMaximal in degree (in-hub) : 84\n",
      "\tRoots num: 179 Leaves num: 201\n",
      "\tSingle Edges num: 1901 Mutual Edges num: 49\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 46496\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.610+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t261\t213.7+-14.7\t3.23\t0.000\t15\t5.61\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "110\t26\t10.7+-2.8\t5.39\t0.000\t7\t0.56\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   34.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Neural_Anterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Neural_Crest_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Neural_Crest_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum 20somite_Neural_Crest network motif computing completed\n",
      "Neural_Posterior\n",
      "of Nodes: 424 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 424\n",
      "\tMaximal out degree (out-hub) : 76\n",
      "\tMaximal in degree (in-hub) : 61\n",
      "\tRoots num: 120 Leaves num: 229\n",
      "\tSingle Edges num: 1907 Mutual Edges num: 45\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 41690\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.655+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1223\t872.8+-40.3\t8.70\t0.000\t30\t29.34\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t355\t212.5+-16.4\t8.68\t0.000\t13\t8.52\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t55\t19.1+-5.0\t7.23\t0.000\t10\t1.32\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t36\t16.2+-3.9\t5.14\t0.000\t6\t0.86\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   33.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Neural_Crest_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Neural_Posterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Neural_Posterior_mfinder_format.txt\n",
      "\tNetwork type: Direct20somite_Neural_Posterior network motif computing completed\n",
      "Notochord\n",
      "ed\n",
      "\tNum of Nodes: 453 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 453\n",
      "\tMaximal out degree (out-hub) : 56\n",
      "\tMaximal in degree (in-hub) : 89\n",
      "\tRoots num: 179 Leaves num: 190\n",
      "\tSingle Edges num: 1879 Mutual Edges num: 59\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 43149\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.643+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 4 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t361\t292.5+-17.0\t4.02\t0.000\t17\t8.37\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "98\t16\t7.0+-3.1\t2.89\t0.010\t6\t0.37\n",
      "\n",
      "0 1 0 \n",
      "0 0 1 \n",
      "1 0 0 \n",
      "\n",
      "108\t25\t12.2+-3.7\t3.48\t0.010\t8\t0.58\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "110\t38\t17.3+-3.8\t5.42\t0.000\t8\t0.88\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   33.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Neural_Posterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Notochord_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     17 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Notochord_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nod20somite_Notochord network motif computing completed\n",
      "PSM\n",
      "es: 477 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 477\n",
      "\tMaximal out degree (out-hub) : 120\n",
      "\tMaximal in degree (in-hub) : 32\n",
      "\tRoots num: 130 Leaves num: 284\n",
      "\tSingle Edges num: 1961 Mutual Edges num: 16\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 42532\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.680+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1100\t624.0+-40.1\t11.86\t0.000\t26\t25.86\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t88\t12.9+-6.0\t12.63\t0.000\t6\t2.07\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t60\t10.2+-3.5\t14.13\t0.000\t6\t1.41\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   34.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Notochord_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_PSM_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     32 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      8 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_PSM_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 296 Num 20somite_PSM network motif computing completed\n",
      "Somites\n",
      "of Edges: 2000\n",
      "\tNum of Nodes with edges: 296\n",
      "\tMaximal out degree (out-hub) : 114\n",
      "\tMaximal in degree (in-hub) : 40\n",
      "\tRoots num: 58 Leaves num: 193\n",
      "\tSingle Edges num: 1923 Mutual Edges num: 38\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 65869\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.358+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 1 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "46\t727\t574.7+-22.5\t6.77\t0.000\t10\t11.04\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   40.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_20somite_PSM_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_20somite_Somites_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     26 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     22 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      7 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      4 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_20somite_Somites_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 20somite_Somites network motif computing completed\n",
      "unassigned\n",
      "465 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 465\n",
      "\tMaximal out degree (out-hub) : 111\n",
      "\tMaximal in degree (in-hub) : 41\n",
      "\tRoots num: 127 Leaves num: 281\n",
      "\tSingle Edges num: 1936 Mutual Edges num: 31\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 43805\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.680+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t1254\t822.9+-47.3\t9.12\t0.000\t26\t28.63\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t285\t175.1+-16.8\t6.55\t0.000\t9\t6.51\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t39\t15.7+-3.7\t6.37\t0.000\t6\t0.89\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   37.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.4 seconds.)\n",
      "\n",
      "Output File motifs_20somite_Somites_OUT.txt was generated\n"
     ]
    }
   ],
   "source": [
    "# Run mFinder for all timepoints and all cell-types\n",
    "\n",
    "links_filepath = \"/hpc/projects/data.science/yangjoon.kim/zebrahub_multiome/data/processed_data/baseGRN_CisBP_RNA_zebrahub/\"\n",
    "\n",
    "for stage in timepoints:\n",
    "    # import the Links object\n",
    "    GRN_stage = co.load_hdf5(links_filepath + \"08_\"+stage+\"_celltype_GRNs.celloracle.links\")\n",
    "    # grab only the filtered GRNs (2000 edges for each GRN)\n",
    "    GRN_all = GRN_stage.filtered_links\n",
    "    \n",
    "    # for each cell-type, run mFinder\n",
    "    for celltype in GRN_all.keys():\n",
    "\n",
    "        print(celltype)\n",
    "\n",
    "        # Step 1. subset the GRN for each celltype\n",
    "        GRN_celltype = GRN_all[celltype]\n",
    "\n",
    "        # Step 2. reformat the GRN (unique integers for gene_names)    \n",
    "        list_genes_TFs = list(set(GRN_celltype.source).union(set(GRN_celltype.target)))\n",
    "        # Create a dictionary mapping integers to gene names\n",
    "        gene_dict = {index: gene_name for index, gene_name in enumerate(list_genes_TFs)}\n",
    "        gene_dict\n",
    "\n",
    "        # Now, we will reformat the GRN as described above\n",
    "        # 1) grab the GRN, then extract the \"source\", \"target\", and create a dataframe\n",
    "        # 2) add the \"edge weight\" as \"1\" for the third column\n",
    "        df_mfinder = pd.DataFrame(columns =[\"source\", \"target\", \"edge_weight\"])\n",
    "        df_mfinder\n",
    "\n",
    "        df_mfinder[\"source\"] = GRN_celltype[\"source\"]\n",
    "        df_mfinder[\"target\"] = GRN_celltype[\"target\"]\n",
    "        df_mfinder[\"edge_weight\"] = 1\n",
    "\n",
    "        df_mfinder\n",
    "\n",
    "        # 3) convert the \"source\", \"target\" gene_names to \"integers\" using the gene_dict\n",
    "        df_mfinder[\"source\"] = df_mfinder[\"source\"].map({v: k for k, v in gene_dict.items()})\n",
    "        df_mfinder[\"target\"] = df_mfinder[\"target\"].map({v: k for k, v in gene_dict.items()})\n",
    "        df_mfinder\n",
    "        # save the reformatted GRN into a txt file\n",
    "        df_mfinder.to_csv(filepath + \"filtered_GRN_\"+stage+\"_\"+celltype+\"_mfinder_format.txt\",\n",
    "                          sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "        # Step 3. run mFinder\n",
    "        # default setting (network_size=3, num_random_)\n",
    "        # input filename\n",
    "        input_file = \"filtered_GRN_\"+stage+\"_\"+celltype+\"_mfinder_format.txt\"\n",
    "        # output filename\n",
    "        output_file = \"motifs_\"+stage+\"_\"+celltype\n",
    "\n",
    "        # define the mFinder command\n",
    "        cmd = \"mfinder \"+input_file + \" -f \"+output_file\n",
    "        cmd\n",
    "        # run mFinder\n",
    "        os.system(cmd)\n",
    "        print(stage + \"_\" + celltype + \" network motif computing completed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "44ec7802-7e4a-47a5-b686-5f5072c7539a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# move to the directory where the output files will be saved\n",
    "filepath = \"/hpc/projects/data.science/yangjoon.kim/zebrahub_multiome/data/processed_data/baseGRN_CisBP_RNA_zebrahub/09_network_motifs/\"\n",
    "os.chdir(filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "667b4f64-4ad0-4366-9480-9f166cf7ab70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the Links object\n",
    "GRN_links_TDR118 = co.load_hdf5(\"/hpc/projects/data.science/yangjoon.kim/zebrahub_multiome/data/processed_data/TDR118_cicero_output/08_TDR118_celltype_GRNs.celloracle.links\")\n",
    "GRN_links_TDR118\n",
    "\n",
    "# grab only the filtered object (2000 edges)\n",
    "GRN_all = GRN_links_TDR118.filtered_links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d993a0e8-2bec-4f4a-bd1a-1a587fe62b3b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'stage' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mstage\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'stage' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "64ef2705-d7c3-457d-b07a-c730b9764bc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adaxial_Cells\n",
      "Input Network file is filtered_GRN_Adaxial_Cells_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Adaxial_Cells_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 455 NuAdaxial_Cells network motif computing completed\n",
      "Differentiating_Neurons\n",
      "m of Edges: 2000\n",
      "\tNum of Nodes with edges: 455\n",
      "\tMaximal out degree (out-hub) : 56\n",
      "\tMaximal in degree (in-hub) : 47\n",
      "\tRoots num: 212 Leaves num: 211\n",
      "\tSingle Edges num: 1973 Mutual Edges num: 11\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 36825\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.720+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 3 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t760\t601.8+-38.9\t4.07\t0.000\t18\t20.64\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t61\t36.7+-5.8\t4.22\t0.000\t5\t1.66\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t26\t14.2+-4.4\t2.66\t0.020\t5\t0.71\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   30.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Adaxial_Cells_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Differentiating_Neurons_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Differentiating_Neurons_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tDifferentiating_Neurons network motif computing completed\n",
      "Endoderm\n",
      "Num of Nodes: 443 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 443\n",
      "\tMaximal out degree (out-hub) : 41\n",
      "\tMaximal in degree (in-hub) : 51\n",
      "\tRoots num: 227 Leaves num: 179\n",
      "\tSingle Edges num: 1973 Mutual Edges num: 11\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 38336\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.692+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 0 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   30.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Differentiating_Neurons_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Endoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Endoderm_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 455 Num of EdgesEndoderm network motif computing completed\n",
      "Epidermal\n",
      ": 2000\n",
      "\tNum of Nodes with edges: 455\n",
      "\tMaximal out degree (out-hub) : 47\n",
      "\tMaximal in degree (in-hub) : 46\n",
      "\tRoots num: 220 Leaves num: 195\n",
      "\tSingle Edges num: 1981 Mutual Edges num: 8\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 39285\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.680+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 0 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   29.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Endoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Epidermal_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     36 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     28 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Epidermal_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 475 Num of EdgEpidermal network motif computing completed\n",
      "Lateral_Mesoderm\n",
      "es: 2000\n",
      "\tNum of Nodes with edges: 475\n",
      "\tMaximal out degree (out-hub) : 63\n",
      "\tMaximal in degree (in-hub) : 50\n",
      "\tRoots num: 198 Leaves num: 237\n",
      "\tSingle Edges num: 1975 Mutual Edges num: 11\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 40218\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.661+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 0 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Epidermal_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Lateral_Mesoderm_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Lateral_Mesoderm_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: Lateral_Mesoderm network motif computing completed\n",
      "Muscle\n",
      "492 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 492\n",
      "\tMaximal out degree (out-hub) : 53\n",
      "\tMaximal in degree (in-hub) : 44\n",
      "\tRoots num: 175 Leaves num: 266\n",
      "\tSingle Edges num: 1970 Mutual Edges num: 13\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 39959\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.669+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 0 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Lateral_Mesoderm_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Muscle_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Muscle_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 491 Num of Edges: 20Muscle network motif computing completed\n",
      "NMPs\n",
      "00\n",
      "\tNum of Nodes with edges: 491\n",
      "\tMaximal out degree (out-hub) : 53\n",
      "\tMaximal in degree (in-hub) : 42\n",
      "\tRoots num: 239 Leaves num: 217\n",
      "\tSingle Edges num: 1978 Mutual Edges num: 10\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 36321\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.729+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 0 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   29.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Muscle_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_NMPs_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_NMPs_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 389 Num of Edges: 2000\n",
      "\tNMPs network motif computing completed\n",
      "Neural_Anterior\n",
      "Num of Nodes with edges: 389\n",
      "\tMaximal out degree (out-hub) : 39\n",
      "\tMaximal in degree (in-hub) : 60\n",
      "\tRoots num: 236 Leaves num: 127\n",
      "\tSingle Edges num: 1959 Mutual Edges num: 15\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 43793\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.625+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t578\t483.3+-29.7\t3.18\t0.000\t13\t13.20\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "108\t23\t6.8+-2.6\t6.34\t0.000\t4\t0.53\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_NMPs_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Neural_Anterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Neural_Anterior_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 42Neural_Anterior network motif computing completed\n",
      "Neural_Crest\n",
      "3 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 423\n",
      "\tMaximal out degree (out-hub) : 46\n",
      "\tMaximal in degree (in-hub) : 49\n",
      "\tRoots num: 183 Leaves num: 197\n",
      "\tSingle Edges num: 1980 Mutual Edges num: 8\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 43052\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.626+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 0 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Neural_Anterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Neural_Crest_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Neural_Crest_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 418 Num Neural_Crest network motif computing completed\n",
      "Neural_Posterior\n",
      "of Edges: 2000\n",
      "\tNum of Nodes with edges: 418\n",
      "\tMaximal out degree (out-hub) : 49\n",
      "\tMaximal in degree (in-hub) : 52\n",
      "\tRoots num: 194 Leaves num: 183\n",
      "\tSingle Edges num: 1972 Mutual Edges num: 12\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 42897\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.632+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 0 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Neural_Crest_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Neural_Posterior_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Neural_Posterior_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: Neural_Posterior network motif computing completed\n",
      "Notochord\n",
      "435 Num of Edges: 2000\n",
      "\tNum of Nodes with edges: 435\n",
      "\tMaximal out degree (out-hub) : 51\n",
      "\tMaximal in degree (in-hub) : 44\n",
      "\tRoots num: 189 Leaves num: 200\n",
      "\tSingle Edges num: 1958 Mutual Edges num: 18\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 40904\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.647+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 1 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "108\t12\t5.5+-2.2\t2.92\t0.010\t4\t0.29\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Neural_Posterior_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Notochord_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     11 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Notochord_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 447 Num of EdgNotochord network motif computing completed\n",
      "PSM\n",
      "es: 2000\n",
      "\tNum of Nodes with edges: 447\n",
      "\tMaximal out degree (out-hub) : 33\n",
      "\tMaximal in degree (in-hub) : 47\n",
      "\tRoots num: 244 Leaves num: 170\n",
      "\tSingle Edges num: 1980 Mutual Edges num: 7\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 36582\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.716+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 2 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "38\t517\t294.1+-23.4\t9.52\t0.000\t19\t14.13\n",
      "\n",
      "0 1 1 \n",
      "0 0 1 \n",
      "0 0 0 \n",
      "\n",
      "46\t19\t11.0+-2.9\t2.77\t0.010\t5\t0.52\n",
      "\n",
      "0 1 1 \n",
      "1 0 1 \n",
      "0 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   28.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Notochord_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_PSM_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     19 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     14 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_PSM_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 448 Num of Edges: 2000\n",
      "\tNuPSM network motif computing completed\n",
      "Somites\n",
      "m of Nodes with edges: 448\n",
      "\tMaximal out degree (out-hub) : 61\n",
      "\tMaximal in degree (in-hub) : 65\n",
      "\tRoots num: 229 Leaves num: 185\n",
      "\tSingle Edges num: 1947 Mutual Edges num: 24\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 41785\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.652+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 1 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "108\t18\t9.4+-3.2\t2.70\t0.010\t4\t0.43\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   32.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    1.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_PSM_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_Somites_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     21 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     18 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     15 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     12 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      9 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_Somites_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 455 Num of Edges: Somites network motif computing completed\n",
      "unassigned\n",
      "2000\n",
      "\tNum of Nodes with edges: 455\n",
      "\tMaximal out degree (out-hub) : 49\n",
      "\tMaximal in degree (in-hub) : 56\n",
      "\tRoots num: 227 Leaves num: 191\n",
      "\tSingle Edges num: 1962 Mutual Edges num: 15\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 41442\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.658+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 1 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "108\t20\t7.2+-2.9\t4.38\t0.000\t5\t0.48\n",
      "\n",
      "0 0 1 \n",
      "1 0 1 \n",
      "1 0 0 \n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   31.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_Somites_OUT.txt was generated\n",
      "Input Network file is filtered_GRN_unassigned_mfinder_format.txt\n",
      "mfinder Version 1.20\n",
      "\n",
      "Loading Network\n",
      "\tReading Network file in <Source,Target,Weight> Format\n",
      "Searching motifs size 3\n",
      "Processing Real network...\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "Processing Random networks\n",
      "..........\n",
      " Estimated run time left :     27 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     24 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     23 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     20 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     16 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     13 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :     10 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      6 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      3 seconds.\n",
      "\n",
      "..........\n",
      " Estimated run time left :      0 seconds.\n",
      "\n",
      "\n",
      "Calculating Results...\n",
      "\n",
      "MOTIF FINDER RESULTS:\n",
      "\n",
      "\tNetwork name: filtered_GRN_unassigned_mfinder_format.txt\n",
      "\tNetwork type: Directed\n",
      "\tNum of Nodes: 392 Num of Eunassigned network motif computing completed\n",
      "dges: 2000\n",
      "\tNum of Nodes with edges: 392\n",
      "\tMaximal out degree (out-hub) : 29\n",
      "\tMaximal in degree (in-hub) : 76\n",
      "\tRoots num: 259 Leaves num: 113\n",
      "\tSingle Edges num: 1973 Mutual Edges num: 5\n",
      "\n",
      "\tMotif size searched  3\n",
      "\tTotal number of 3-node subgraphs : 50843\n",
      "\tNumber of random networks generated : 100\n",
      "\tRandom networks generation method: Switches\n",
      "\tNum of Switches range: 100.0-200.0, Success switches Ratio:0.601+-0.00\n",
      "\n",
      "The following motifs were found:\n",
      "\n",
      "Criteria taken : Nreal Zscore > 2.00\n",
      "                 Pval ignored (due to small number of random networks)\n",
      "                 Mfactor > 1.10\n",
      "                 Uniqueness >= 4\n",
      "\n",
      "\n",
      "\n",
      "\tFull list includes 0 motifs\n",
      "\n",
      "MOTIF\tNREAL\tNRAND\t\tNREAL\tNREAL\tUNIQ\tCREAL\n",
      "ID\t\tSTATS\t\tZSCORE\tPVAL\tVAL\t[MILI]\t\n",
      "\n",
      "\n",
      "\n",
      " (Application total runtime was:   32.0 seconds.)\n",
      "\n",
      " (Real network processing runtime was:    0.0 seconds.)\n",
      "\n",
      " (Single Random network processing runtime was:    0.3 seconds.)\n",
      "\n",
      "Output File motifs_unassigned_OUT.txt was generated\n"
     ]
    }
   ],
   "source": [
    "# Run mFinder for each timepoint, for each celltype, then save the result to the \"filepath\"\n",
    "\n",
    "for timepoints in stages:\n",
    "\n",
    "    for celltype in GRN_all.keys():\n",
    "\n",
    "        print(celltype)\n",
    "\n",
    "        # Step 1. subset the GRN for each celltype\n",
    "        GRN_celltype = GRN_all[celltype]\n",
    "\n",
    "        # Step 2. reformat the GRN (unique integers for gene_names)    \n",
    "        list_genes_TFs = list(set(GRN_celltype.source).union(set(GRN_celltype.target)))\n",
    "        # Create a dictionary mapping integers to gene names\n",
    "        gene_dict = {index: gene_name for index, gene_name in enumerate(list_genes_TFs)}\n",
    "        gene_dict\n",
    "\n",
    "        # Now, we will reformat the GRN as described above\n",
    "        # 1) grab the GRN, then extract the \"source\", \"target\", and create a dataframe\n",
    "        # 2) add the \"edge weight\" as \"1\" for the third column\n",
    "        df_mfinder = pd.DataFrame(columns =[\"source\", \"target\", \"edge_weight\"])\n",
    "        df_mfinder\n",
    "\n",
    "        df_mfinder[\"source\"] = GRN_celltype[\"source\"]\n",
    "        df_mfinder[\"target\"] = GRN_celltype[\"target\"]\n",
    "        df_mfinder[\"edge_weight\"] = 1\n",
    "\n",
    "        df_mfinder\n",
    "\n",
    "        # 3) convert the \"source\", \"target\" gene_names to \"integers\" using the gene_dict\n",
    "        df_mfinder[\"source\"] = df_mfinder[\"source\"].map({v: k for k, v in gene_dict.items()})\n",
    "        df_mfinder[\"target\"] = df_mfinder[\"target\"].map({v: k for k, v in gene_dict.items()})\n",
    "        df_mfinder\n",
    "        # save the reformatted GRN into a txt file\n",
    "        df_mfinder.to_csv(filepath + \"filtered_GRN_\"+celltype+\"_mfinder_format.txt\",\n",
    "                          sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "        # Step 3. run mFinder\n",
    "        # default setting (network_size=3, num_random_)\n",
    "        # input filename\n",
    "        input_file = \"filtered_GRN_\"+celltype+\"_mfinder_format.txt\"\n",
    "        # output filename\n",
    "        output_file = \"motifs_\" + celltype\n",
    "\n",
    "        # define the mFinder command\n",
    "        cmd = \"mfinder \"+input_file + \" -f \"+output_file\n",
    "        cmd\n",
    "        # run mFinder\n",
    "        os.system(cmd)\n",
    "        print(celltype + \" network motif computing completed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eceb341c-a14d-4beb-aab4-f5320d6720ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for celltype in GRN_all.keys():\n",
    "    \n",
    "#     print(celltype)\n",
    "    \n",
    "#     # Step 3. run mFinder\n",
    "#     # default setting (network_size=3, num_random_)\n",
    "#     # input filename\n",
    "#     input_file = \"filtered_GRN_\"+celltype+\"_mfinder_format.txt\"\n",
    "#     # output filename\n",
    "#     output_file = \"motifs_\" + celltype\n",
    "\n",
    "#     # define the mFinder command\n",
    "#     cmd = \"mfinder \"+input_file + \" -f \"+output_file\n",
    "#     cmd\n",
    "#     # run mFinder\n",
    "#     os.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff502a23-80c4-4858-9737-dc3f3af8ff34",
   "metadata": {},
   "source": [
    "NOTES\n",
    "- Refer to mFinder documentation (Uri Alon's lab website)\n",
    "- link: \n",
    "\n",
    "1) Use \"screen\"\n",
    "2) the motif computation takes less than a minute (with the default settings, network_size=3, for a GRN with 2000 edges, from CellOracle).\n",
    "3) however, if we increase the network_size=4, the runtime increased to 20 minutes, for the same input dataset.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbc03fd0-7018-407b-ae60-c2bb3a1954eb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a246e895-d376-4d27-8f83-347902369a45",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f54abd-aae7-453a-9316-b0a7f20c1d59",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e6aea82-5729-46ac-9e25-cabbfca81a4a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-celloracle_env]",
   "language": "python",
   "name": "conda-env-.conda-celloracle_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
